File: frontend/src/types/review.ts
/**
 * Review Types
 * TypeScript interfaces for the review and rating system
 */

export interface ReviewCreate {
  ride_id: string;
  reviewee_id: string;
  rating: number;
  comment?: string;
}

export interface ReviewerInfo {
  id: string;
  full_name: string;
  avatar_url?: string;
}

export interface Review {
  id: string;
  ride_id: string;
  reviewer_id: string;
  reviewee_id: string;
  rating: number;
  comment?: string;
  created_at: string;
}

export interface ReviewWithReviewer {
  id: string;
  ride_id: string;
  rating: number;
  comment?: string;
  created_at: string;
  reviewer: ReviewerInfo;
}

export interface PaginatedReviews {
  reviews: ReviewWithReviewer[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface ReviewStats {
  rating_avg: number;
  rating_count: number;
  five_star_count: number;
  four_star_count: number;
  three_star_count: number;
  two_star_count: number;
  one_star_count: number;
}

File: backend/src/schemas/review.py
"""
Review API Schemas
Pydantic models for review and rating endpoints.
"""
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, field_validator
from uuid import UUID


# ===== REQUEST SCHEMAS =====

class ReviewCreate(BaseModel):
    """Schema for creating a new review"""
    ride_id: UUID
    reviewee_id: UUID
    rating: int
    comment: Optional[str] = None
    
    @field_validator('rating')
    def validate_rating(cls, v):
        if v < 1 or v > 5:
            raise ValueError('Rating must be between 1 and 5')
        return v
    
    @field_validator('comment')
    def validate_comment(cls, v):
        if v is not None:
            v = v.strip()
            if len(v) == 0:
                return None  # Empty comment treated as None
            if len(v) > 150:
                raise ValueError('Comment cannot exceed 150 characters')
        return v


# ===== RESPONSE SCHEMAS =====

class ReviewerInfo(BaseModel):
    """Minimal reviewer information for review responses"""
    id: UUID
    full_name: str
    avatar_url: Optional[str] = None
    
    class Config:
        from_attributes = True


class ReviewResponse(BaseModel):
    """Schema for review response"""
    id: UUID
    ride_id: UUID
    reviewer_id: UUID
    reviewee_id: UUID
    rating: int
    comment: Optional[str] = None
    created_at: datetime
    
    class Config:
        from_attributes = True


class ReviewWithReviewer(BaseModel):
    """Schema for review with reviewer details"""
    id: UUID
    ride_id: UUID
    rating: int
    comment: Optional[str] = None
    created_at: datetime
    reviewer: ReviewerInfo
    
    class Config:
        from_attributes = True


class PaginatedReviewsResponse(BaseModel):
    """Schema for paginated reviews list"""
    reviews: list[ReviewWithReviewer]
    total: int
    page: int
    page_size: int
    total_pages: int


class ReviewStatsResponse(BaseModel):
    """Schema for review statistics"""
    rating_avg: float
    rating_count: int
    five_star_count: int
    four_star_count: int
    three_star_count: int
    two_star_count: int
    one_star_count: int
